### PR TITLE
Add command for copying jack-in command to clipboard

### LINKED ISSUE #992: Add command for copying jack-in command to clipboard
I've done this process several times, especially when I'm working on a Calva feature that requires the repl and I want to leave the repl running in a terminal external to VS Code.

1. Start jack-in
2. Copy the command
3. Kill the jack-in terminal process

It would be nice if I could just run a command that copies the jack-in command that would be run to my clipboard, instead of having to do the steps above. I don't think the command needs a default keybinding. I'd be fine just running it from the command palette.

### PR DESCRIPTION
**Pull request by bpringe**
## What has Changed?

- Adds command to copy jack-in command (that would be executed in a terminal) to clipboard
- Refactors the jack-in code in order to make this work

@PEZ This will need a review and testing, of course.

Closes #992 

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [ ] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. NB: *There is a CircleCI bug that makes the Artifacts hard to find. Please see [this issue](https://discuss.circleci.com/t/artifacts-tab-not-showing-unless-logged-in/32433) for workarounds.*
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [ ] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [ ] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [ ] Created the issue I am fixing/addressing, if it was not present.
- [ ] Added to or updated docs in this branch, if appropriate

## The Calva Team PR Checklist:

Before merging we (at least one of us) have:

- [ ] Made sure the PR is directed at the `dev` branch (unless reasons).
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and tested it there if so.
- [ ] Read the source changes.
- [ ] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @kstehn, @cfehse, @bpringe

### CODE REVIEW

**Comment by bpringe** [comment]
@pez If you could test the jack-in env thing that would be helpful. I did slightly change the code for that but it should do the same thing.

File: src/nrepl/jack-in.ts [hunk 1]
```diff
@@ -108,6 +101,70 @@ export function calvaJackout() {
     liveShareSupport.didJackOut();
 }
 
+export async function copyJackInCommandToClipboard(): Promise<void> {
+    try {
+        await state.initProjectDir();
+    } catch (e) {
+        console.error('An error occurred while initializing project directory.', e);
+        return;
+    }
```
Comment (bpringe) [inline src/nrepl/jack-in.ts:109]: This is done both here and in the jack-in process. Is there a reason to do it when jack-in is run and not earlier, like when Calva is activated?

Comment (PEZ) [reply src/nrepl/jack-in.ts:109]: I don't quite recall, but I think it makes sense when you have several projects in the window. Then which one will be jacked-in to will be determined by the active editor, which we don't really know about at startup.

**Comment by PEZ** [comment]
jackInEnv is still working, afaict:

```json
    "calva.jackInEnv": {
        "foo": 100
    }
```

Jack-in!

```clojure
(System/getenv "foo") => "100"
```

**Review by PEZ — APPROVED** [review]
Looks good. I renamed a function whose name confused me.

The only thing to consider is if we want it to copy out a complete command line, including the jackInEnv. Don't know how to do that on Windows, and it might bring us in to that quoting hell that is so fun. Maybe it can remain a TODO. =)

**Comment by bpringe** [comment]
Adding a note here for myself to add back analytics that were removed, as per our convo outside of here.

**Comment by bpringe** [comment]
> The only thing to consider is if we want it to copy out a complete command line, including the jackInEnv. 

I don't think I fully understand what you mean, but as far as I can tell we currently are copying to clipboard everything that would be printed/run in the terminal at jack-in. Maybe leaving it as a todo would be good.

**Comment by PEZ** [comment]
Say we the user has a `jackInEnv` setting an environment of `foo=bar`. Then the jack-in command-line is not enough for recreating the jack-in. But something like

```sh
foo=bar clojure -Sdeps '{:deps {nrepl/nrepl {:mvn/version "0.8.3"} cider/cider-nrepl {:mvn/version "0.25.8"}}}' -A:reveal
```

would.

Anyway. I'm totally fine with merging as it is now.

**Comment by PEZ** [comment]
Merged as I am working at another jack-in thing and don't want to stitch it together later. 😄

**Comment by bpringe** [comment]
You still want me to add back analytics that were removed?

**Comment by PEZ** [comment]
Maybe you can add that back in #1004? Since it refactors some of the jack-in further...

**Comment by bpringe** [comment]
I'll leave it as a todo (#1002) until the dust settles on all the current jack-in work.

**Pull request merged**